nemo/collections/tts/models/spectrogram_enhancer.py

```diff
@@ -0,0 +1,382 @@
+            first_part = [torch.randn(batch_size, self._cfg["latent_dim"])] * mixing_point
+            second_part = [torch.randn(batch_size, self._cfg["latent_dim"])] * (self.G.num_layers - mixing_point)
+            zs = [*first_part, *second_part]
+        else:
+            zs = [torch.randn(batch_size, self._cfg["latent_dim"])] * self.G.num_layers
+
+        return self.move_to_correct_device(zs)
+
+    def generate_noise(self, batch_size: int = 1) -> Tensor:
+        noise = torch.rand(batch_size, self._cfg["n_bands"], 4096, 1)
+        return self.move_to_correct_device(noise)
+
+    def pad_spectrogram(self, spectrogram):
+        multiplier = 2 ** sum(1 for block in self.G.blocks if block.upsample)
+        *_, max_length = spectrogram.shape
+        return torch.nn.functional.pad(spectrogram, (0, multiplier - max_length % multiplier))
+
+    def forward(
+        self,
+        condition: Tensor,
+        lengths: Tensor,
+        zs: Optional[List[Tensor]] = None,
+        ws: Optional[List[Tensor]] = None,
+        noise: Optional[Tensor] = None,
+        mixing: bool = False,
+        normalize: bool = True,
+    ):
+        if len(condition.shape) != 4:
+            raise ValueError(f"Got spectrogram tensor of shape {condition.shape}, expected B x 1 x C x L")
+
+        batch_size, *_, max_length = condition.shape
+
+        # generate noise
+        if zs is not None and ws is not None:
+            raise ValueError(
+                "Please specify either zs or ws or neither, but not both. It is not clear which one to use."
+            )
+
+        if zs is None:
+            zs = self.generate_zs(batch_size, mixing)
+        if ws is None:
+            ws = [self.G.style_mapping(z) for z in zs]
+        if noise is None:
+            noise = self.generate_noise(batch_size)
+
+        # normalize if needed, mask and pad appropriately
+        if normalize:
+            condition = self.normalize_spectrograms(condition)
+        condition = self.pad_spectrogram(condition)
+        condition = mask(condition, lengths)
+
+        # the main call
+        enhanced_spectrograms = self.G(condition, lengths, ws, noise)
+
+        # denormalize if needed, mask and remove padding
+        if normalize:
+            enhanced_spectrograms = self.unnormalize_spectrograms(enhanced_spectrograms)
+        enhanced_spectrograms = mask(enhanced_spectrograms, lengths)
+        enhanced_spectrograms = enhanced_spectrograms[:, :, :, :max_length]
+
+        return enhanced_spectrograms
+
+    def prepare_batch(self, batch):
+        if not isinstance(self.spectrogram_model, FastPitchModel):
+            raise NotImplementedError(
+                f"{type(self.spectrogram_model)} is not supported, please implement batch preparation for this model."
+            )
+
+        attn_prior, durs, speaker = None, None, None
+        (audio, audio_lens, text, text_lens, attn_prior, pitch, _, speaker,) = batch
```

Comment:
Use this function instead for uniformity: https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/tts/helpers/helpers.py#L564